### PR TITLE
Strictly increasing check with test coverage for streamplot grid

### LINKED ISSUE #18898: wrong bounds checking in streamplot start_points
### Bug report

The bounds check for the `start_points` argument is wrong if the `x` or `y` coordinates are not in ascending order:

```python
import matplotlib.pylab as plt
import numpy as np

u = np.ones((2, 2))
v = np.zeros((2, 2))
x = np.array([0, 10])

y = np.array([10, 0])  # (a)
#y = np.array([0, 10])  # (b)

plt.streamplot(x, y, u, v, start_points=[[5, 5]])  # (c)
#plt.streamplot(x, y, u, v)  # (d)

plt.show()
```

The code results in
```
ValueError: Starting point (5.0, 5.0) outside of data boundaries
```

I would have expected that a plot with one streamline is created. The error goes away **either** if I replace `(a)` with `(b)` **or** if I replace `(c)` with `(d)`. Thus if `start_points` are **not** given, plotting a descending range works totally fine.

**Matplotlib version**

  * Operating system: macOS 10.15.7
  * Matplotlib version: 3.3.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 3.8.6
  * Other libraries: numpy: 1.19.2

Matplotlib has been installed via `pip`.

### PR DESCRIPTION
**Pull request by aitikgupta**
## PR Summary
This PR fixes #18898, adds check for strictly increasing `x` and `y` in `matplotlib.streamplot.Grid`, and adds test coverage for it.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/tests/test_streamplot.py [hunk 1]
```diff
@@ -114,3 +115,53 @@ def test_streamplot_limits():
...
+def test_streamplot_grid():
+    u = np.ones((2, 2))
+    v = np.zeros((2, 2))
+
+    # Test for same rows and columns
+    x = np.array([[10, 20], [10, 30]])
+    y = np.array([[10, 10], [20, 20]])
+
+    msg = "The rows of 'x' must be equal"
+    with pytest.raises(ValueError, match=msg):
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_streamplot.py:129]: I'd rather inline the message in the function. Having a separate variable is an unnecessary indirection.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/tests/test_streamplot.py (hunk 1 above)
Comment (aitikgupta) [reply lib/matplotlib/tests/test_streamplot.py]: Something like I just pushed?

File: lib/matplotlib/tests/test_streamplot.py [hunk 2]
```diff
@@ -114,3 +115,49 @@ def test_streamplot_limits():
...
+
+    x = np.array([[10, 20], [10, 20]])
+    y = np.array([[10, 10], [20, 20]])
+    plt.streamplot(x, y, u, v)
+
+    # Test for maximum dimensions
+    x = np.array([0, 10])
+    y = np.array([[[0, 10]]])
+
+    with pytest.raises(ValueError, match="'y' can have at maximum"
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_streamplot.py:145]:
```suggestion
    with pytest.raises(ValueError, match="'y' can have at maximum "
```

Comment (aitikgupta) [reply lib/matplotlib/tests/test_streamplot.py]: 👍🏼

**Comment by tacaswell** [comment]
Thanks @aitikgupta !  Congratulations on your first Matplotlib PR :tada: , hopefully we will hear from you again!

**Pull request merged**